File: src/data/tauri/dto.ts
/**
 * Tauri wire 格式(snake_case)。
 * 只有 mappers.ts 可以 import 這個檔;app/ui 層永遠看不到 snake_case。
 */

export type TaskStatusDto = "draft" | "ready" | "completed";

export interface TaskDto {
  id: string;
  title: string;
  target_date: string;
  created_at: string;
  updated_at: string;
  sort_order: number;
  status: TaskStatusDto;
  completed_at: string | null;
  difficulty_selected: number | null;
  difficulty_suggested: "easy" | "medium" | "hard" | null;
  base_points: number;
  final_reward_points: number;
  is_focus_task: boolean;
}

export interface WalletDto {
  wallet_points: number;
  pending_today_points: number;
  streak_days: number;
  last_completed_date: string | null;
  last_rollover_date: string | null;
  best_streak_days: number;
  lifetime_points_earned: number;
  points_spent_on_unlocks: number;
  pending_today_unlock_spent: number;
}

export interface DailySummaryDto {
  date: string;
  tasks_created: number;
  tasks_completed: number;
  focus_tasks_completed: number;
  points_earned: number;
  is_all_completed: boolean;
}

export interface SettingsDto {
  theme_mode: "light" | "dark" | "system";
  panel_width: number;
  pin_enabled: boolean;
  refresh_time: string;
  trigger_position_y: number;
  locale: string;
}

export interface UnlockStatusDto {
  max_visible_task_slots: number;
  focus_task_feature_unlocked: boolean;
  custom_refresh_time_unlocked: boolean;
  weekly_analysis_unlocked: boolean;
}

export interface UnlockNodeStateDto {
  node_id: string;
  category: string;
  cost: number;
  requires: string[];
  unlocked: boolean;
  unlocked_at: string | null;
  requirements_met: boolean;
  affordable: boolean;
}

export interface UnlockProgressDto {
  available_points: number;
  lifetime_points_earned: number;
  points_spent_on_unlocks: number;
  nodes: UnlockNodeStateDto[];
}

export interface BootstrapResponseDto {
  today: string;
  tasks: TaskDto[];
  wallet: WalletDto;
  settings: SettingsDto;
  summary: DailySummaryDto | null;
  unlocks: UnlockStatusDto;
  refresh_applied: boolean;
}

export interface CompleteTaskResponseDto {
  task: TaskDto;
  reward_earned: number;
  bonus_earned: number;
  pending_today_points: number;
  wallet_points: number;
  today_summary: DailySummaryDto;
  all_tasks_completed: boolean;
  streak_days: number;
  available_points_delta: number;
  available_points_after: number;
}

export interface PurchaseUnlockResponseDto {
  node_id: string;
  unlocks: UnlockStatusDto;
  available_points: number;
  points_spent_on_unlocks: number;
  pending_today_unlock_spent: number;
}

export interface RefreshResponseDto {
  refresh_applied: boolean;
  previous_date: string | null;
  new_date: string;
  wallet: WalletDto;
  previous_summary: DailySummaryDto | null;
}

export interface SettingsPatchDto {
  theme_mode?: "light" | "dark" | "system";
  panel_width?: number;
  pin_enabled?: boolean;
  refresh_time?: string;
  locale?: string;
}

export interface CommandErrorDto {
  code: string;
  message: string;
  details?: Record<string, unknown>;
}

File: apps/desktop/src-tauri/src/dto.rs
//! Wire 回應格式(序列化給前端的 response envelope)。
//! 欄位名與舊版 API 相容,前端 Phase 1 只需處理 Task.status 與結構化錯誤兩個變化。

use crate::catalog::unlock::UnlockStatus;
use crate::domain::settings::Settings;
use crate::domain::summary::DailySummary;
use crate::domain::task::Task;
use crate::domain::wallet::Wallet;
use serde::Serialize;

#[derive(Debug, Serialize)]
pub struct BootstrapResponse {
    pub today: String,
    pub tasks: Vec<Task>,
    pub wallet: Wallet,
    pub settings: Settings,
    pub summary: Option<DailySummary>,
    pub unlocks: UnlockStatus,
    pub refresh_applied: bool,
}

#[derive(Debug, Serialize)]
pub struct CompleteTaskResponse {
    pub task: Task,
    pub reward_earned: i64,
    pub bonus_earned: i64,
    pub pending_today_points: i64,
    pub wallet_points: i64,
    pub today_summary: DailySummary,
    pub all_tasks_completed: bool,
    pub streak_days: i64,
    pub available_points_delta: i64,
    pub available_points_after: i64,
}

#[derive(Debug, Serialize)]
pub struct RefreshResponse {
    pub refresh_applied: bool,
    pub previous_date: Option<String>,
    pub new_date: String,
    pub wallet: Wallet,
    pub previous_summary: Option<DailySummary>,
}

#[derive(Debug, Serialize)]
pub struct UnlockNodeState {
    pub node_id: String,
    pub category: String,
    pub cost: i64,
    pub requires: Vec<String>,
    pub unlocked: bool,
    pub unlocked_at: Option<String>,
    pub requirements_met: bool,
    pub affordable: bool,
}

#[derive(Debug, Serialize)]
pub struct UnlockProgressResponse {
    pub available_points: i64,
    pub lifetime_points_earned: i64,
    pub points_spent_on_unlocks: i64,
    pub nodes: Vec<UnlockNodeState>,
}

#[derive(Debug, Serialize)]
pub struct PurchaseUnlockResponse {
    pub node_id: String,
    pub unlocks: UnlockStatus,
    pub available_points: i64,
    pub points_spent_on_unlocks: i64,
    pub pending_today_unlock_spent: i64,
}
